### ISSUE TITLE
Global plugins are not applied to discriminators

### ISSUE THREAD

**Issue by alexis-y** [Automattic/mongoose#4965]
Per the docs:

> Want to register a plugin for all schemas? The mongoose singleton has a .plugin() function that registers a plugin for every schema.

However, it doesn't apply it to discriminator schemas.

```JavaScript
const assert = require('assert');
const { Mongoose, Schema } = require('mongoose');
const sinon = require('sinon');

describe('global plugins', () => {
  it('should be applied to discriminators', () => {
    let mongoose = new Mongoose();
    let plugin = sinon.spy();
    mongoose.plugin(plugin);

    let shapeSchema = new Schema({});
    let Shape = mongoose.model('Shape', shapeSchema);

    let circleSchema = new Schema({});
    let Circle = Shape.discriminator('Circle', circleSchema);

    assert(plugin.withArgs(shapeSchema).calledOnce,
      'applied once to the base schema');  // Ok
    assert(plugin.withArgs(circleSchema).calledOnce,
      'applied once to the discriminator schema'); // Mmm, nope
    assert(plugin.calledTwice,
      'applied twice overall');  // Also nope

  });
});
```

I think either the plugins should be applied to the discriminator schemas as well (which I imagine is a breaking change with many unintended consequences), or it should be called out in the docs.

**Version tested**
node 6.7.0
mongoose 4.7.9

**Comment by varunjayaraman** [comment]
@alexis-y sweet, thanks for the repro script. Looks like this is a bug since global plugins should definitely be applied to all schemas.

**Comment by vkarpov15** [comment]
Good call. I'm probably going to have to put this behind a flag because the last time we changed `mongoose.plugin()` there were a lot of issues

**Comment by sebastian-nowak** [comment]
It's too bad it's not covered in the docs.

For anyone wondering how to use this, you have to add:

```js
mongoose.set('applyPluginsToDiscriminators', true);
```

**Comment by vkarpov15** [comment]
Will add to docs, thanks for pointing out this gap

**Comment by chumager** [comment]
Hi, I've a discriminator and the plugins doesn't work, maybe I'm doing something wrong. 

`const muv = require("mongoose-unique-validator");
const {Schema} = mongoose;
//mongoose.set("applyPluginsToDiscriminators", true);
mongoose.connect("mongodb://localhost/test");
mongoose.plugin(muv);
let test = new Schema({
    _id:Number,
});
let test2 = new Schema({
    b:String,
});
let Test = mongoose.model("Test",test);
let Test2 = Test.discriminator("Test2",test2);
console.log(Test.schema.path("_id"));
console.log(Test2.schema.path("_id"));
Test.create({_id:1}).catch(err=>console.log("TEST",err));
Test2.create({_id:1}).catch(err=>console.log("TEST2",err));
`
In both models there is the same validator, but It is not applied in the discriminator model.
It happened with others plugins even my own ones, any clues???

**Comment by vkarpov15** [comment]
Thanks for reporting, will investigate asap

**Comment by vkarpov15** [comment]
@chumager not sure what you mean by "not applied in the discriminator model". You're not even using mongoose-unique-validator in your script, you're just applying the plugin but not declaring any paths as `unique`. The below script works as expected:

```javascript
const mongoose = require('mongoose');
const muv = require("mongoose-unique-validator");
const {Schema} = mongoose;

mongoose.set("applyPluginsToDiscriminators", true);
mongoose.connect("mongodb://localhost/test");
mongoose.plugin(muv);
let test = new Schema({
    _id:Number,
});
let test2 = new Schema({
    b:{type: String, unique: true }
});
let Test = mongoose.model("Test",test);
let Test2 = Test.discriminator("Test2",test2);
console.log(Test.schema.path("_id"));
console.log(Test2.schema.path("b"));
```

**Comment by chumager** [comment]
@vkarpov15 hi, the plugin works with unique and _id index, it was just an example, I've a very complex Model creation and plugins that's why I used this module as an example, but if you check my example, the parent schema uses the muv and send a parsed error, but the discriminated model does not.

the console log shows different errors... 
TEST { ValidationError: Test validation failed: _id: Error, expected `_id` to be unique. Value: `1`
TEST2 { MongoError: E11000 duplicate key error index: test.tests.$_id_ dup key: { : 1 }

the problem does not exists when you apply the plugin to the discriminator, my problem es with the global plugins, you can define the unique (even though the plugin works with _id index) in a parent key and the result will be the same.

```
"use strict";
const mongoose = require("mongoose");
const muv = require("mongoose-unique-validator");
const {Schema} = mongoose;
mongoose.set("applyPluginsToDiscriminators", true);
mongoose.connect("mongodb://localhost/test");
mongoose.plugin(muv);
let test = new Schema({
    a:{
        type:Number,
        unique:true
    }
});
let test2 = new Schema({
    b:String,
});
let Test = mongoose.model("Test",test);
let Test2 = Test.discriminator("Test2",test2);
console.log(Test.schema.path("a"));
console.log(Test2.schema.path("a"));
Test.create([{a:1},{a:1}]).catch(err=>console.log("TEST",err));
Test2.create([{a:1},{a:1}]).catch(err=>console.log("TEST2",err));

```
Look at the path, both has the same validator (muv uses a validator to check for data to be unique), then the error in TEST works ok, but not in TEST2.

TEST { ValidationError: Test validation failed: a: Error, expected `a` to be unique. Value: `1`
TEST2 { MongoError: E11000 duplicate key error index: test.tests.$a_1 dup key: { : 1 }

Do I make myself clear? my problem is when I use a global plugin, the discriminator doesn't use it in the inherited paths. With or without applyPluginsToDiscriminators 

After analysing the similarities with my plugins and muv plugin, I think the problem could be not in the plugin itself but the way is constructed. We both create conditions travelling through paths. 

I think that's why it's not applied to discriminators even with inherited paths like _id.

I hope this could help.

**Comment by chumager** [comment]
Hi, I've found another example in my plugin code.

```
            if(!schema.methods.checkPasswd){
                let checkPasswd;
                if(schema.path("password")){
                    checkPasswd = function(password){
                        return bcrypt.compare(password,this.password);
                    };
                }else{
                    checkPasswd = ()=>{
                        return Q(false);
                    };
                }
                schema.method({checkPasswd});
            }

```

As the path password does not exist in the discriminator schema, but in the parent schema, the result is always a false promise.

This problem just happened few days ago, maybe because in my connections pool set I setted applyPluginsToDiscriminators into the db connectors (new mongoose.Mongoose), instead of mongoose.

to fix that I've changed the behavior so I can use **this** 

```
            if(!schema.methods.checkPasswd){
                schema.method({
                    checkPasswd(password){
                        if(this.schema.path("password")){
                            Log.silly("checkPasswd","existe el path",this.schema.path("password"));
                            return bcrypt.compare(password,this.password);
                        }else{
                            Log.silly("checkPasswd","no existe el path");
                            return Q(false);
                        }
                    }
                });
            }

```

So... Even if the static, method or hook is available in the parent, it's not applied to the discriminator.

For this particular problem I could fixed it, but there other cases when I've not access to **this** so I can't fix it.

Maybe it's a order problem, when the global plugin is applied to the discriminator, I don't know.

Regards.

**Comment by vkarpov15** [comment]
Confirmed, below script reproduces the issue:

```javascript
"use strict";
const mongoose = require("mongoose");
const muv = require("mongoose-unique-validator");
const {Schema} = mongoose;
mongoose.set("applyPluginsToDiscriminators", true);
mongoose.connect("mongodb://localhost/test");
mongoose.plugin(muv);
let test = new Schema({
    a:{
        type:Number,
        unique:true
    }
});
let test2 = new Schema({
    b:String,
});
let Test = mongoose.model("Test",test);
let Test2 = Test.discriminator("Test2",test2);
console.log(Test.schema.path("a"));
console.log(Test2.schema.path("a"));
Test.create([{a:1},{a:1}]).catch(err=>console.log("TEST",err));
Test2.create([{a:1},{a:1}]).catch(err=>console.log("TEST2",err));
```

Will fix ASAP

**Comment by vsDizzy** [comment]
Still not fixed in **5.6.8**.
Any progress/workaround on this?

**Comment by vkarpov15** [comment]
@vsDizzy the issue from [my last comment](https://github.com/Automattic/mongoose/issues/4965#issuecomment-406417373) was a bug in mongoose-unique-validator that has since been fixed. If your issue is with mongoose-unique-validator, please upgrade to the latest version of mongoose-unique-validator. Otherwise, please take a look at [the `applyPluginsToDiscriminators` option](https://mongoosejs.com/docs/api.html#mongoose_Mongoose-set)
